vcd/resource_vcd_vapp_vm.go

```diff
@@ -1919,32 +2152,61 @@ func genericVcdVmRead(d *schema.ResourceData, meta interface{}, origin string) d
 	dSet(d, "cpu_hot_add_enabled", vm.VM.VMCapabilities.CPUHotAddEnabled)
 	dSet(d, "memory_hot_add_enabled", vm.VM.VMCapabilities.MemoryHotAddEnabled)
 
-	if vm.VM.VmSpecSection != nil && vm.VM.VmSpecSection.MemoryResourceMb != nil {
-		dSet(d, "memory", vm.VM.VmSpecSection.MemoryResourceMb.Configured)
-		dSet(d, "memory_priority", vm.VM.VmSpecSection.MemoryResourceMb.SharesLevel)
-		if vm.VM.VmSpecSection.MemoryResourceMb.Reservation != nil {
-			dSet(d, "memory_reservation", vm.VM.VmSpecSection.MemoryResourceMb.Reservation)
-		}
-		if vm.VM.VmSpecSection.MemoryResourceMb.Limit != nil {
-			dSet(d, "memory_limit", vm.VM.VmSpecSection.MemoryResourceMb.Limit)
+	if vm.VM.VmSpecSection != nil {
+		if vm.VM.VmSpecSection.MemoryResourceMb != nil {
+			dSet(d, "memory", vm.VM.VmSpecSection.MemoryResourceMb.Configured)
+			dSet(d, "memory_priority", vm.VM.VmSpecSection.MemoryResourceMb.SharesLevel)
+			if vm.VM.VmSpecSection.MemoryResourceMb.Reservation != nil {
+				dSet(d, "memory_reservation", vm.VM.VmSpecSection.MemoryResourceMb.Reservation)
+			}
+			if vm.VM.VmSpecSection.MemoryResourceMb.Limit != nil {
+				dSet(d, "memory_limit", vm.VM.VmSpecSection.MemoryResourceMb.Limit)
+			}
+			if vm.VM.VmSpecSection.MemoryResourceMb.Shares != nil {
+				dSet(d, "memory_shares", vm.VM.VmSpecSection.MemoryResourceMb.Shares)
+			}
 		}
-		if vm.VM.VmSpecSection.MemoryResourceMb.Shares != nil {
-			dSet(d, "memory_shares", vm.VM.VmSpecSection.MemoryResourceMb.Shares)
+		dSet(d, "cpus", vm.VM.VmSpecSection.NumCpus)
+		dSet(d, "cpu_cores", vm.VM.VmSpecSection.NumCoresPerSocket)
+		if vm.VM.VmSpecSection.CpuResourceMhz != nil {
+			if vm.VM.VmSpecSection.CpuResourceMhz.Reservation != nil {
+				dSet(d, "cpu_reservation", vm.VM.VmSpecSection.CpuResourceMhz.Reservation)
+			}
+			if vm.VM.VmSpecSection.CpuResourceMhz.Limit != nil {
+				dSet(d, "cpu_limit", vm.VM.VmSpecSection.CpuResourceMhz.Limit)
+			}
+			if vm.VM.VmSpecSection.CpuResourceMhz.Shares != nil {
+				dSet(d, "cpu_shares", vm.VM.VmSpecSection.CpuResourceMhz.Shares)
+			}
+			dSet(d, "cpu_priority", vm.VM.VmSpecSection.CpuResourceMhz.SharesLevel)
 		}
-	}
-	dSet(d, "cpus", vm.VM.VmSpecSection.NumCpus)
-	dSet(d, "cpu_cores", vm.VM.VmSpecSection.NumCoresPerSocket)
-	if vm.VM.VmSpecSection != nil && vm.VM.VmSpecSection.CpuResourceMhz != nil {
-		if vm.VM.VmSpecSection.CpuResourceMhz.Reservation != nil {
-			dSet(d, "cpu_reservation", vm.VM.VmSpecSection.CpuResourceMhz.Reservation)
+
+		if vm.VM.VmSpecSection.HardwareVersion != nil {
+			dSet(d, "hardware_version", vm.VM.VmSpecSection.HardwareVersion.Value)
 		}
-		if vm.VM.VmSpecSection.CpuResourceMhz.Limit != nil {
-			dSet(d, "cpu_limit", vm.VM.VmSpecSection.CpuResourceMhz.Limit)
+		if vm.VM.VmSpecSection.OsType != "" {
+			dSet(d, "os_type", vm.VM.VmSpecSection.OsType)
 		}
-		if vm.VM.VmSpecSection.CpuResourceMhz.Shares != nil {
-			dSet(d, "cpu_shares", vm.VM.VmSpecSection.CpuResourceMhz.Shares)
+
+		dSet(d, "firmware", vm.VM.VmSpecSection.Firmware)
```

Comment:
Just double checking, as all over dSets have an nil/empty "if" upfront. It's not needed here? And this line is not under the new bootOptions block below on purpose?

Reply:
We already check if `vmSpecSection` is not set to nil on line 2155. If it's not nil, then `vm.VM.VmSpecSection.Firmware` is initialized its zero value (`""`), so we can set it without checking.

 I'm not sure why there are other checks for strings as it doesn't really make a difference, only for *int or *boot types